newIDE/app/src/MainFrame/UseAuthenticatedPlayer.js

```diff
@@ -0,0 +1,95 @@
+// @flow
+import * as React from 'react';
+import AuthenticatedUserContext from '../Profile/AuthenticatedUserContext';
+import { getPlayerToken } from '../Utils/GDevelopServices/Play';
+import PreferencesContext from './Preferences/PreferencesContext';
+
+const gd: libGDevelop = global.gd;
+
+type PlayerTokensDict = { [gameId: string]: string };
+
+type UseAuthenticatedPlayerOutput = {|
+  getAuthenticatedPlayerForPreview: (
+    currentProject: ?gdProject
+  ) => Promise<?{|
+    playerId: string,
+    playerUsername: string,
+    playerToken: string,
+  |}>,
+|};
+
+export const useAuthenticatedPlayer = (): UseAuthenticatedPlayerOutput => {
+  const { profile, getAuthorizationHeader } = React.useContext(
+    AuthenticatedUserContext
+  );
+  const { values: preferencesValues } = React.useContext(PreferencesContext);
+  const playerTokensForPreview = React.useRef<PlayerTokensDict>({});
+
+  const getAuthenticatedPlayerForPreview = React.useCallback(
+    async (currentProject: ?gdProject) => {
+      if (
+        !profile ||
+        !currentProject ||
+        !preferencesValues.fetchPlayerTokenForPreviewAutomatically
+      )
+        return null;
+
+      const gameId = currentProject.getProjectUuid();
+
+      const playerTokenForPreview = playerTokensForPreview.current[gameId];
+      if (playerTokenForPreview) {
+        // If we already have a token, let's use it.
+        return {
+          playerId: profile.id,
+          playerUsername: profile.username || 'Player',
+          playerToken: playerTokenForPreview,
+        };
+      }
+
+      const isMultiplayerOrPlayerAuthenticationExtensionUsed = gd.UsedExtensionsFinder.scanProject(
+        currentProject
+      )
+        .getUsedExtensions()
+        .toNewVectorString()
+        .toJSArray()
+        .some(
+          extensionName =>
+            extensionName === 'Multiplayer' ||
+            extensionName === 'PlayerAuthentication'
+        );
+      if (!isMultiplayerOrPlayerAuthenticationExtensionUsed) return;
+
+      const newPlayerTokenForGame = await getPlayerToken({
```

Comment:
What if this fails? Should there be a retry?

Reply:
good question, I haven't thought of that option. I'll add something